plugins/node/instrumentation-kafkajs/test/kafkajs.test.ts

```diff
@@ -0,0 +1,941 @@
+
+const instrumentation = registerInstrumentationTesting(
+  new KafkaJsInstrumentation()
+);
+
+import * as kafkajs from 'kafkajs';
+import {
+  Kafka,
+  ProducerRecord,
+  RecordMetadata,
+  Producer,
+  ProducerBatch,
+  Message,
+  Consumer,
+  ConsumerRunConfig,
+  EachBatchPayload,
+  EachMessagePayload,
+  KafkaMessage,
+} from 'kafkajs';
+import { DummyPropagation } from './DummyPropagation';
+
+describe('instrumentation-kafkajs', () => {
+  propagation.setGlobalPropagator(
+    new CompositePropagator({
+      propagators: [new DummyPropagation(), new W3CBaggagePropagator()],
+    })
+  );
+
+  const kafka = new Kafka({
+    clientId: 'unit-tests',
+    brokers: ['testing_mock_host:1234'],
+  });
+
+  let producer: Producer;
+  let messagesSent: Message[] = [];
+
+  const patchProducerSend = (cb: () => Promise<RecordMetadata[]>) => {
+    const origProducerFactory = kafkajs.Kafka.prototype.producer;
+    kafkajs.Kafka.prototype.producer = function (...args): Producer {
+      const producer = origProducerFactory.apply(this, args);
+
+      producer.send = function (record: ProducerRecord) {
+        messagesSent.push(...record.messages);
+        return cb();
+      };
+
+      producer.sendBatch = function (batch: ProducerBatch) {
+        batch.topicMessages?.forEach(topicMessages =>
+          messagesSent.push(...topicMessages.messages)
+        );
+        return cb();
+      };
+
+      return producer;
+    };
+  };
+
+  let consumer: Consumer;
+  let runConfig: ConsumerRunConfig | undefined;
+
+  const storeRunConfig = () => {
+    const origConsumerFactory = kafkajs.Kafka.prototype.consumer;
+    kafkajs.Kafka.prototype.consumer = function (...args): Consumer {
+      const consumer: Consumer = origConsumerFactory.apply(this, args);
+      consumer.run = function (config?: ConsumerRunConfig): Promise<void> {
+        runConfig = config;
+        return Promise.resolve();
+      };
+      return consumer;
+    };
+  };
+
+  beforeEach(() => {
+    messagesSent = [];
+  });
+
+  describe('producer', () => {
+    const expectKafkaHeadersToMatchSpanContext = (
+      kafkaMessage: Message,
+      span: ReadableSpan
+    ) => {
+      assert.strictEqual(
+        kafkaMessage.headers?.[DummyPropagation.TRACE_CONTEXT_KEY],
+        span.spanContext().traceId
+      );
+      assert.strictEqual(
+        kafkaMessage.headers?.[DummyPropagation.SPAN_CONTEXT_KEY],
+        span.spanContext().spanId
+      );
+    };
+
+    describe('successful send', () => {
+      beforeEach(async () => {
+        patchProducerSend(async (): Promise<RecordMetadata[]> => {
+          return [
+            {
+              topicName: 'topic-name-1',
+              partition: 0,
+              errorCode: 123,
+              offset: '18',
+              timestamp: '123456',
+            },
+          ];
+        });
+        instrumentation.disable();
+        instrumentation.enable();
+        producer = kafka.producer();
```

Comment:
Producer here does not send messages to any Kafka instance but rather keeps them in a memory. There are other instrumentations (ioredis, mongodb, mysql, ...) that use docker to get closer to a real use case scenario. They are slower but IMO there is value on it. What do you think?

cc: @pichlermarc @blumamir

Reply:
I think writing the tests with a real kafka broker instance can improve the tests and catch more issues, but it is also a lot of work. If there is motivation for someone to work on it, I support and encourage. If the task is too time-consuming or complicated, I am also good with keeping the tests as they are now.

Reply:
let's keep it put of the scope of this PR :)

I'll try to allocate som time for it and create a PR for that. Thanks @blumamir